test/TokenSale.spec.ts

```diff
@@ -0,0 +1,249 @@
+import {expect} from 'chai';
+import chai from 'chai';
+import hre, {ethers, waffle} from 'hardhat';
+import {IERC20, TokenLock, TokenSale} from '../typechain';
+import {ZERO_ADDRESS, BN, MAX_UINT} from './shared/Constants';
+import {BigNumber} from 'ethers';
+
+const {solidity, loadFixture} = waffle;
+chai.use(solidity);
+
+let tokenIn: IERC20;
+let tokenOut: IERC20;
+let tokenLock: TokenLock;
+let tokenSale: TokenSale;
+const SALE_START = Math.floor(new Date(`2021-12-24T12:00:00.000Z`).getTime() / 1000);
+const ONE_DAY = 24 * 60 * 60;
+const ARENA_DECIMALS = 18;
+const ONE_18 = BigNumber.from(`10`).pow(`18`);
+const ONE_ARENA = ethers.utils.parseUnits(`1`, ARENA_DECIMALS);
+// Polygon USDC only has 6 decimals
+const TOKEN_IN_DECIMALS = 6;
+const ONE_USDC = ethers.utils.parseUnits(`1`, TOKEN_IN_DECIMALS);
+// 0.05 USDC per 1 ARENA (0.05e6 * 1e18 / 1*1e18)
+const TOKEN_OUT_PRICE = ethers.utils.parseUnits(`0.05`, TOKEN_IN_DECIMALS).mul(ONE_18).div(ONE_ARENA);
+const SALE_DURATION = 7 * ONE_DAY;
+const WHITELISTED_ACCOUNTS: string[] = [];
+const WHITELISTED_AMOUNTS = [BN.from(`1000`).mul(ONE_ARENA), BN.from(`500`).mul(ONE_ARENA)];
+
+describe('TokenSale', async () => {
+  const [user, admin, saleRecipient, buyer1, buyer2, other] = waffle.provider.getWallets();
+  WHITELISTED_ACCOUNTS.push(buyer1.address, buyer2.address);
+
+  async function fixture() {
+    let TokenFactory = await ethers.getContractFactory('TestERC20');
+    let tokenIn = (await TokenFactory.connect(admin).deploy('USDC', 'USDC')) as IERC20;
+    let tokenOut = (await TokenFactory.connect(admin).deploy('ARENA', 'ARENA')) as IERC20;
+    let TokenLockFactory = await ethers.getContractFactory('TokenLock');
```

Comment:
instead of using `TokenLock` in deployment it would be better to use `RevokableTokenLock` contract

Reply:
fixed